src/Microsoft.CodeQuality.Analyzers.Exp/Core/Usage/DisposableFieldsShouldBeDisposed.cs

```diff
@@ -0,0 +1,203 @@
+
+        internal static DiagnosticDescriptor Rule = new DiagnosticDescriptor(RuleId,
+                                                                             s_localizableTitle,
+                                                                             s_localizableMessage,
+                                                                             DiagnosticCategory.Usage,
+                                                                             DiagnosticHelpers.DefaultDiagnosticSeverity,
+                                                                             isEnabledByDefault: DiagnosticHelpers.EnabledByDefaultIfNotBuildingVSIX,
+                                                                             description: s_localizableDescription,
+                                                                             helpLinkUri: "https://docs.microsoft.com/en-us/visualstudio/code-quality/ca2213-disposable-fields-should-be-disposed",
+                                                                             customTags: WellKnownDiagnosticTags.Telemetry);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            context.RegisterCompilationStartAction(compilationContext =>
+            {
+                if (!DisposeAnalysisHelper.TryGetOrCreate(compilationContext.Compilation, out DisposeAnalysisHelper disposeAnalysisHelper))
+                {
+                    return;
+                }
+
+                var fieldDisposeValueMap = new ConcurrentDictionary<IFieldSymbol, /*disposed*/bool>();
+                Action<IFieldSymbol, bool> addOrUpdateFieldDisposedValue = (field, disposed) =>
+                {
+                    Debug.Assert(!field.IsStatic);
+                    Debug.Assert(field.Type.IsDisposable(disposeAnalysisHelper.IDisposable));
+
+                    fieldDisposeValueMap.AddOrUpdate(field,
+                        addValue: disposed,
+                        updateValueFactory: (f, currentValue) => currentValue || disposed);
+                };
+
+                // Disposable fields with initializer at declaration must be disposed.
+                compilationContext.RegisterOperationAction(operationContext =>
+                {
+                    var initializedFields = ((IFieldInitializerOperation)operationContext.Operation).InitializedFields;
+                    foreach (var field in initializedFields)
+                    {
+                        if (!field.IsStatic &&
+                            disposeAnalysisHelper.GetDisposableFields(field.ContainingType).Contains(field))
+                        {
+                            addOrUpdateFieldDisposedValue(field, /*disposed*/false);
```

Comment:
Prefer named parameters so that if this is ever refactored, it will be updated appropriately.